src/test/completions_test.ts

```diff
@@ -0,0 +1,280 @@
+      if (!editor || !editor.shadowRoot) return reject();
+      const focusContainer = editor.shadowRoot.querySelector('#focusContainer');
+      if (!focusContainer) return reject();
+
+      const config = {childList: true};
+      // To avoid computer/dom specific timing errors in tests, we rely on mutations
+      const observer = new MutationObserver(async (mutationsList, obs) => {
+        if (addedNodesContainsCompletionsMenu(mutationsList)) {
+          obs.disconnect();
+          await waitMilli(100);
+          resolve('');
+        }
+      });
+
+      if (focusContainer) {
+        observer.observe(focusContainer, config);
+      }
+
+      setTimeout(() => {
+        observer.disconnect();
+        resolve('');
+      }, 10000);
+    });
+  const waitForCompletionsToDisappear = () =>
+    new Promise((resolve, reject) => {
+      // Make sure we can grab the focuscontainer for observing
+      if (!editor || !editor.shadowRoot) return reject();
+      const focusContainer = editor.shadowRoot.querySelector('#focusContainer');
+      if (!focusContainer) return reject();
+
+      const config = {childList: true};
+      // To avoid computer/dom specific timing errors in tests, we rely on mutations
+      const observer = new MutationObserver(async (mutationsList, obs) => {
+        if (removedNodesContainsCompletionsMenu(mutationsList)) {
+          obs.disconnect();
+          await waitMilli(100);
+          resolve('');
+        }
+      });
+
+      if (focusContainer) {
+        observer.observe(focusContainer, config);
+      }
+      setTimeout(() => {
+        observer.disconnect();
+        resolve('');
+      }, 10000);
+    });
+  const addedNodesContainsCompletionsMenu = (mutationsList: MutationRecord[]) =>
+    mutationsList.some((mut) =>
+      Array.from(mut.addedNodes).some((node) =>
+        (node as Element).classList.contains('CodeMirror-hints')
+      )
+    );
+  const removedNodesContainsCompletionsMenu = (
+    mutationsList: MutationRecord[]
+  ) =>
+    mutationsList.some((mut) =>
+      Array.from(mut.removedNodes).some((node) =>
+        (node as Element).classList.contains('CodeMirror-hints')
+      )
+    );
+  const raf = async () => new Promise((r) => requestAnimationFrame(r));
+  const waitMilli = (ms: number) => new Promise((r) => setTimeout(r, ms));
+  const pierce = async (...selectors: string[]) => {
+    let node = document.body;
+    for (const selector of selectors) {
+      const result = (node.shadowRoot ?? node).querySelector(selector);
+      assert.instanceOf(result, Element);
+      if ((result as ReactiveElement).updateComplete) {
+        await (result as ReactiveElement).updateComplete;
+      }
+      node = result as HTMLElement;
+    }
+    return node;
+  };
+  const waitForCompileDone = () =>
+    new Promise((resolve) => {
+      project?.addEventListener(
+        'compileDone',
+        async () => {
+          waitMilli(1000);
```

Comment:
Are these extra waits because there's no explicit signal we can listen to that we've finished rendering completion UI after each change? Maybe another approach could be to poll every N milliseconds until the DOM we expect has appeared, timing out after M seconds?